sensor_msgs/include/sensor_msgs/field_generators.hpp

```diff
@@ -0,0 +1,170 @@
+/// has_member_x<PointXYZ>::value
+#define IMPL__LIDAR_UTILS__DEFINE_MEMBER_CHECKER(MEMBER) \
+  template<typename T, typename V = bool> \
+  struct has_member_ ## MEMBER : std::false_type {}; \
+  template<typename T> \
+  struct has_member_ ## MEMBER< \
+    T, typename std::enable_if< \
+      !std::is_same<decltype(std::declval<T>().MEMBER), void>::value, \
+      bool>::type>: std::true_type {}
+
+#define IMPL__LIDAR_UTILS__DEFINE_MEMBER_FUNCTION_CHECKER(FUNCTION_NAME) \
+  template<typename T, typename = void> \
+  struct has_function_ ## FUNCTION_NAME ## _returning_non_const_ref : std::false_type {}; \
+  template<typename T> \
+  struct has_function_ ## FUNCTION_NAME ## _returning_non_const_ref< \
+    T, std::enable_if_t< \
+      std::is_reference<decltype(std::declval<T>().FUNCTION_NAME())>::value && \
+      !std::is_const<std::remove_reference_t<decltype( \
+        std::declval<T>().FUNCTION_NAME())>>::value>>: std::true_type {}
+
+/// A Macro that defines a generator struct for a given field name.
+///
+/// The generator structs take a vector of PointFields and push_back a new
+/// PointField into it if the underlying class has the specified field. The
+/// existence of the field is checked by the has_member_FIELD<PointT>::value.
+#define LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(FIELD) \
+  struct field_ ## FIELD ## _generator { \
+    IMPL__LIDAR_UTILS__DEFINE_MEMBER_CHECKER(FIELD); \
+    IMPL__LIDAR_UTILS__DEFINE_MEMBER_FUNCTION_CHECKER(FIELD); \
+    template<class PointT> \
+    static inline std::enable_if_t< \
+      !has_member_ ## FIELD<PointT>::value && \
+      !has_function_ ## FIELD ## _returning_non_const_ref<PointT>::value> \
+    push_back_field_if_needed(sensor_msgs::msg::PointCloud2::_fields_type &) { \
+    } \
+    template<class PointT> \
+    static inline std::enable_if_t< \
+      has_member_ ## FIELD<PointT>::value || \
+      has_function_ ## FIELD ## _returning_non_const_ref<PointT>::value> \
+    push_back_field_if_needed(sensor_msgs::msg::PointCloud2::_fields_type & fields) { \
+      sensor_msgs::detail::push_back_field(fields, #FIELD, &PointT::FIELD); \
+    } \
+  }
+
+LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(x);
+LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(y);
+LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(z);
+LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(id);
+LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(ring);
+LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER(intensity);
+
+/// A help function to start iterating over all field adders and return the appropritate fields
+/// generated by those.
+///
+/// @tparam     PointT       Type of point
+/// @tparam     FieldAdders  A tuple of field adder types, generated by
+///                          LIDAR_UTILS__DEFINE_FIELD_GENERATOR_FOR_MEMBER
+///
+/// @return     All generated PointField structs.
+///
+template<class PointT, class FieldAdders>
+inline sensor_msgs::msg::PointCloud2::_fields_type generate_fields_from_point()
```

Comment:
A conversation starter: this is the meat of the field generator. This calls a recursive function that generates all the fields that are needed from the point structure. Is this clear enough? Can this be improved? Any thoughts are welcome.